File: rx-fitness/src/main/java/com/nicolasmilliard/rxfit/SensorObservable.java
package com.nicolasmilliard.rxfit;

import com.google.android.gms.fitness.SensorsClient;
import com.google.android.gms.fitness.data.DataPoint;
import com.google.android.gms.fitness.request.OnDataPointListener;
import com.google.android.gms.fitness.request.SensorRequest;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.nicolasmilliard.rxtask.ObservableTask;
import com.nicolasmilliard.rxtask.ObservableTaskCallback;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.exceptions.CompositeException;
import io.reactivex.exceptions.Exceptions;
import io.reactivex.plugins.RxJavaPlugins;

final class SensorObservable extends ObservableTask<DataPoint> {

    private final SensorsClient client;
    private final SensorRequest request;

    public SensorObservable(SensorsClient client, SensorRequest request) {
        this.client = client;
        this.request = request;
    }

    @Override
    protected Task<Void> run(ObservableTaskCallback<DataPoint> callback) {
        OnDataPointListener listener = dataPoint -> callback.onNext(dataPoint);
        callback.setDisposeListener(() -> this.client.remove(listener));
        return client.add(this.request, listener)
                .addOnCompleteListener(callback);
    }
}

File: rx-fit/src/main/java/com/nicolasmilliard/rxfit/SensorObservable.kt
package com.nicolasmilliard.rxfit

import com.google.android.gms.fitness.SensorsClient
import com.google.android.gms.fitness.data.DataPoint
import com.google.android.gms.fitness.request.OnDataPointListener
import com.google.android.gms.fitness.request.SensorRequest
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.exceptions.CompositeException
import io.reactivex.exceptions.Exceptions
import io.reactivex.plugins.RxJavaPlugins

class SensorObservable(private val client: SensorsClient, private val request: SensorRequest) : Observable<DataPoint>() {
    override fun subscribeActual(observer: Observer<in DataPoint>) {
        val callback = SensorCallback(client, observer)
        observer.onSubscribe(callback)
        client.add(request, callback).addOnCompleteListener(callback)
    }

    internal class SensorCallback(private val client: SensorsClient, private val observer: Observer<in DataPoint>) : Disposable, OnCompleteListener<Void>, OnDataPointListener {
        override fun onDataPoint(dataPoint: DataPoint) {
            if (disposed) return
            observer.onNext(dataPoint)
        }

        private var disposed = false

        override fun onComplete(task: Task<Void>) {
            if (disposed) return
            if (!task.isSuccessful) {
                try {
                    observer.onError(task.exception!!)
                } catch (inner: Throwable) {
                    Exceptions.throwIfFatal(inner)
                    RxJavaPlugins.onError(CompositeException(task.exception, inner))
                }
            }
        }

        override fun isDisposed(): Boolean = disposed

        override fun dispose() {
            disposed = true
            client.remove(this).addOnCompleteListener({
                // Not sure what to do here
            })
        }

    }
}